Sources/System/FilePath/FilePathSyntax.swift

```diff
@@ -0,0 +1,353 @@
+  /// but not for concat... Rust's `push` does replacement. Should we just do push/pop?
+  ///
+  public mutating func append(_ other: FilePath) {
+    // TODO: Faster to do byte copy, checking for leading separator
+    self.components.append(contentsOf: other.components)
+  }
+
+  /// Remove and return the last component of this file path. If the path is
+  /// root or empty, does nothing and returns `nil`.
+  ///
+  /// Examples:
+  /// * `"/".pop() == nil          // path is /`
+  /// * `"/foo/bar".pop() == "bar" // path is /foo`
+  public mutating func pop() -> FilePath.Component? {
+    self.components.popLast() // FIXME: popLastButNotRoot()
+  }
+
+  /// See `append`
+  ///
+  /// TODO: Should this be `/`? Should the RHS have to be Component?
+  public static func +(_ lhs: FilePath, _ rhs: FilePath) -> FilePath {
+    var result = lhs
+    result.append(rhs)
+    return result
+  }
+
+  /// Collapse "." and ".." components syntactically, that is, without following symlinks
+  ///
+  /// TODO: Should we have a var as well?
+  mutating func lexicallyNormalize() {
+    fatalError()
+  }
+
+  /// Returns `self` in a form that's relative to `base`.
+  /// This does not cosult the file system or resolve symlinks.
+  ///
+  /// Returns `nil` if they do not share the same root
+  ///
+  /// TODO: examples
+  ///
+  /// TODO: C++ adds in the "if any `file_name` in either `relative_path` can be interpreted as a root name, return empty".
+  /// What should we say/do?
+  public func lexicallyRelative(toBase base: FilePath) -> FilePath? {
+    fatalError()
+  }
+
+
+  /// If `prefix` is a prefix of `self`, removes it and returns `true`. Otherwise
+  /// returns `false`.
+  ///
+  /// TODO: var too?
+  public mutating func stripPrefix(_ prefix: FilePath) -> Bool {
+    fatalError()
+  }
+}
+
+// Convenience helpers
+extension FilePath {
+  /// TODO: Should we have this? It is massively annoying
+  /// to have to call String(decoding: path), especially since so
+  /// many of our operations return `FilePath`.
+  public var string: String {
+    String(decoding: self)
+  }
+
+  /// TODO: This would otherwise be pretty annoying for users to do
+  /// themselves
+  public var componentStrings: [String] {
+    self.components.map { String(decoding: $0) }
+  }
+}
+
+// REVIEW NOTE: This is a mock-up of the kind of thing we could provide on Windows
+// as future work. Remove this before merging.
+#if os(Windows)
+extension FilePath {
+  /*public*/ struct WindowsRootName {
+    /*public*/ enum Kind {
+      case drive(Character)
+      case unc(host: String, share: String)
+      case volume(String) // Would this be for \\?\Foo
+    }
+    /*public*/ var kind: Kind { fatalError() }
+    /*public*/ var isVerbatim: Bool { fatalError() }
+  }
+  /*
+   DOS: letter ':'
+   UNC: '\\' host name '\' (share name | letter '$')
+   DOS device:
+   ('\\.\' | '\\?\') (letter ':' | 'Volume{' GUID '}')
+   */
+  /*public*/ var windowsRootName: WindowsRootName? { nil }
```

Comment:
Will there also be a way to get at the various roots on Windows, e.g. a static function to return the root for a particular drive letter on Windows?  Or would that always be done by just creating a particular path with a `C:\` string passed in?

Reply:
Windows can use the `root` member to get the root out, which would be the whole root (ala `C#`'s semantics). I do want to provide further root analysis, akin to what Rust's [Prefix](https://doc.rust-lang.org/std/path/enum.Prefix.html) type can do, though I'm more inclined to go with naming and structure that would be familiar to Windows users over Rust's naming. I have some corner cases to figure out there as well, so I need to run some experiments.

FWIW, what is currently proposed is as much as C++ or C# gives users, I just would like to give even more (especially since I have to write the root parser anyways).

Reply:
What I meant by my question was whether on Windows, I could create a path that had the root corresponding to a particular drive letter (such as one entered by a user) without already having a `FilePath` instance from which to get a root.